### PR TITLE
solution

### PR DESCRIPTION
**Pull request by yelnikov-andrii**

### CODE REVIEW

**Review by alexandra-protyanova — CHANGES_REQUESTED** [review]

File: src/averageAges.js [hunk 1]
```diff
@@ -15,11 +15,25 @@
  * @return {number}
  */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter(person => person['sex'] === 'm');
```
Comment (Anna-Tkachenko) [inline src/averageAges.js:18]: Don't repeat yourself: just check at the start of the function if century argument exists
if yes - you should create the array with men for correct century
if no - you should create the array with all men

and then logic for both of variants is the same: reducing and getting the average value
```suggestion
  const men = century
    ? people.filter(...)
    : people.filter(person => person['sex'] === 'm');
```

File: src/averageAges.js [hunk 2]
```diff
@@ -37,7 +51,28 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter(x => x['sex'] === 'f');
```
Comment (Anna-Tkachenko) [inline src/averageAges.js:54]: Same here, use ternary operator, find correct array of people and call reduce for it
```suggestion
  const women = withChildren
    ? people.filter(...)
    : people.filter(x => x['sex'] === 'f');
```

**Review by Anna-Tkachenko — CHANGES_REQUESTED** [review]

File: src/averageAges.js [hunk 3]
```diff
@@ -15,11 +15,21 @@
...
-  // without nesting
+  let amountOfexistence = 0;
+
+  const men = century
+    ? people.filter(person => Math.ceil(person['died'] / 100)
+    === century && person['sex'] === 'm')
+
+    : people.filter(person => person['sex'] === 'm');
+
+  amountOfexistence = men.reduce((sum, manData) => (manData['died']
```
Comment (Anna-Tkachenko) [inline src/averageAges.js:26]: It's not amountOfExistence, it's sum of ages)

File: src/averageAges.js [hunk 4]
```diff
@@ -37,7 +47,23 @@ function calculateMenAverageAge(people, century) {
...
+  let amountOfexistence = 0;
+
+  const women = withChildren
+    ? people.filter((person) => {
+      const peopleHaveMother = people.map(persone => persone['mother']);
+
+      return peopleHaveMother.includes(person['name']);
+    }) : people.filter(x => x['sex'] === 'f');
+
+  amountOfexistence = women.reduce((sum, woman) =>
```
Comment (Anna-Tkachenko) [inline src/averageAges.js:59]: Same here with variable name

File: src/averageAges.js [hunk 5]
```diff
@@ -55,7 +81,61 @@ function calculateWomenAverageAge(people, withChildren) {
...
+      if (mother !== undefined) {
+        sum += child['born'] - mother['born'];
+        countOfpeople++;
+      }
+
+      return sum;
+    }, 0);
+
+    result = amountOfDifferAges / countOfpeople;
+  }
```
Comment (Anna-Tkachenko) [inline src/averageAges.js:111]: Your this logic are the same for both of variants: if you have onlyWithSon argument and if not
Move this check at the start of the function as you did for previous functions
Just check at the start, if this argument is passed, you should find all kids that are boys
If this argument isn't passes - you should find all kids
And then logic is the same (reduce and calculating average age)

**Review by Anna-Tkachenko — CHANGES_REQUESTED** [review]

File: src/averageAges.js [hunk 6]
```diff
@@ -37,7 +47,23 @@ function calculateMenAverageAge(people, century) {
...
+
+  const women = withChildren
+    ? people.filter((person) => {
+      const peopleHaveMother = people.map(persone => persone['mother']);
+
+      return peopleHaveMother.includes(person['name']);
+    }) : people.filter(x => x['sex'] === 'f');
+
+  sumOfAges = women.reduce((sum, woman) =>
+    (woman['died'] - woman['born'] + sum), 0);
```
Comment (Denchepornyuk) [inline src/averageAges.js:60]: you have two identical reduce, create a function for them

File: src/averageAges.js [hunk 7]
```diff
@@ -55,7 +81,38 @@ function calculateWomenAverageAge(people, withChildren) {
...
+  const amountOfDifferAges = children.reduce((previousValue, child) => {
+    const mother = mothers.find((mom) => {
+      if (mom['name'] === child['mother']) {
+        return mom;
+      }
+    });
+
+    let sum = previousValue;
+
+    if (mother !== undefined) {
```
Comment (Denchepornyuk) [inline src/averageAges.js:105]:
```suggestion
    if (mother) {
```

File: src/averageAges.js [hunk 8]
```diff
@@ -55,7 +81,38 @@ function calculateWomenAverageAge(people, withChildren) {
...
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothersArr = people.map(person => person['mother']);
+  const mothers = people.filter(person => mothersArr.includes(person['name']));
+  const men = people.filter(person => person['sex'] === 'm');
+  let countOfpeople = 0;
+  let result = 0;
+
+  const children = onlyWithSon
+    ? men.filter(y => y['mother']).filter(person => person !== null)
```
Comment (Denchepornyuk) [inline src/averageAges.js:91]: you can get key from object by name, fix in all places
```suggestion
    ? men.filter(y => y.mother).filter(person => person !== null)
```

**Review by denys-cheporniuk — CHANGES_REQUESTED** [review]

**Review by denys-cheporniuk — APPROVED** [review]

**Pull request closed without merging**